src/UseCase/BannerSelection/CampaignConfigurationLoader.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\BannerServer\UseCase\BannerSelection;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Yaml\Exception\ParseException;
+use Symfony\Component\Yaml\Yaml;
+use WMDE\BannerServer\Entity\BannerSelection\Banner;
+use WMDE\BannerServer\Entity\BannerSelection\Bucket;
+use WMDE\BannerServer\Entity\BannerSelection\Campaign;
+use WMDE\BannerServer\Entity\BannerSelection\CampaignCollection;
+use WMDE\BannerServer\Utils\SystemRandomIntegerGenerator;
+
+/**
+ * @license GNU GPL v2+
+ */
+class CampaignConfigurationLoader {
+
+	private const DEFAULT_CONFIGURATION_FILE_PATH = 'config/campaigns/campaigns.yml';
+
+	private $configFile;
+	private $logger;
+
+	public function __construct( LoggerInterface $logger, string $configFile = self::DEFAULT_CONFIGURATION_FILE_PATH ) {
+		$this->configFile = $this->getAbsolutePath( $configFile );
+		$this->logger = $logger;
+	}
+
+	public function getCampaignCollection(): CampaignCollection {
+		$campaigns = [];
+		foreach ( $this->parseConfiguration() as $campaignName => $campaignData ) {
+			$campaign = $this->buildCampaignFromData( $campaignName, $campaignData );
+			if ( $campaign ) {
+				$campaigns[] = $campaign;
+			}
+		}
+		return new CampaignCollection( ...$campaigns );
+	}
+
+	private function buildCampaignFromData( string $campaignName, array $campaignData ): ?Campaign {
+		$buckets = $this->buildBucketsFromData( $campaignData );
+		if ( empty( $buckets ) ) {
+			return null;
+		}
+		return new Campaign(
+			$campaignName,
+			new \DateTime( $campaignData['start'] ),
+			new \DateTime( $campaignData['end'] ),
+			(int)$campaignData['trafficLimit'],
+			new SystemRandomIntegerGenerator(),
+			array_shift( $buckets ),
+			...$buckets
+		);
+	}
+
+	private function buildBucketsFromData( array $campaignData ): array {
+		$buckets = [];
+		foreach ( $campaignData['buckets'] as $bucketData ) {
+			$bucket = $this->buildBucketFromData( $bucketData );
+			if ( $bucket ) {
+				$buckets[] = $bucket;
+			}
+		}
+		return $buckets;
+	}
+
+	private function buildBucketFromData( array $bucketData ): ?Bucket {
+		if ( !isset( $bucketData['name'] ) ) {
+			return null;
+		}
+		$banners = $this->buildBannersFromData( $bucketData['banners'] );
+		if ( empty( $banners ) ) {
+			return null;
+		}
+		return new Bucket( $bucketData['name'], array_shift( $banners ), ...$banners );
+	}
+
+	private function buildBannersFromData( array $bannerData ): array {
+		$banners = [];
+		foreach ( $bannerData as $bannerIdentifier ) {
+			$banners[] = new Banner( $bannerIdentifier );
+		}
+		return $banners;
+	}
+
+	private function parseConfiguration(): array {
+		try {
+			return Yaml::parseFile( $this->configFile );
+		}
+		catch ( ParseException $exception ) {
+			$this->logger->critical( 'Unable to read banner server config file.', [ 'exception' => $exception ] );
+			return [];
+		}
+	}
+
+	private function getAbsolutePath( string $path ): string {
```

Comment:
This could be viewed as a "separation of concerns" smell, because the rest of the class is about parsing and not about locating the config file. I'd suggest removing the default constant and `getAbsolutePath` and putting the path to the config file in the application configuration (as a parameter), with different paths to the different files for each environment.